### PR TITLE
Rspec tests stage 1

### PR DESCRIPTION
**Pull request by middaysan**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -62,9 +62,14 @@ end
 group :test do
   # Adds support for Capybara system testing and selenium driver
   gem 'capybara', '>= 2.15'
+  gem 'rails-controller-testing'
   gem 'selenium-webdriver'
+  gem 'simplecov'
   # Easy installation and use of chromedriver to run system tests with Chrome
   gem 'chromedriver-helper'
+  %w[rspec-core rspec-expectations rspec-mocks rspec-rails rspec-support].each do |lib|
```
Comment (RGBD) [inline Gemfile]: Remind me why did you need to install github version and not rubygems.org?
also, you have git_source helper defined in line 4, google how to use it.
Also never specify master branch as it may be unstable.
Also use git repositories carefully because of security implications.

File: app/controllers/projects_controller.rb [hunk 2]
```diff
@@ -21,12 +21,13 @@ def create
     if @project.save
       redirect_to projects_path
     else
-      render 'new'
+      redirect_to new_project_path
```
Comment (RGBD) [inline app/controllers/projects_controller.rb]: should be render 'new', undo this change
TODO: discuss

File: spec/controllers/profile_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      expect(response).to render_template('edit_role')
+    end
+
+    it 'customer setup info' do
+      @user.role = :developer
+      @user.save!
+      sign_in @user
+      get :customer_setup_info
+      expect(response).to render_template('customer_setup_info')
+    end
```
Comment (RGBD) [inline spec/controllers/profile_controller_spec.rb]: newline below

File: spec/controllers/profile_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
+require 'rails_helper'
+
+RSpec.describe ProfileController, type: :controller do
+  before(:each) do
+    @request.host = '127.0.0.1:3000'
+    @user = User.create!(name: 'midd', email: 'user@example.org', password: 'very-secret')
+  end
+
+  context 'not get view' do
```
Comment (RGBD) [inline spec/controllers/profile_controller_spec.rb]: name is misleading

File: spec/controllers/profile_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    it 'edit role' do
+      sign_in @user
+      get :edit_role
+      expect(response).to render_template('edit_role')
+    end
+
+    it 'customer setup info' do
+      @user.role = :developer
```
Comment (RGBD) [inline spec/controllers/profile_controller_spec.rb]: user.update!(role: :developer)

File: spec/controllers/profile_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      @user.save!
+      sign_in @user
+      get :customer_setup_info
+      expect(response).to render_template('customer_setup_info')
+    end
+    it 'edit' do
+      sign_in @user
+      @user.role = :developer
```
Comment (RGBD) [inline spec/controllers/profile_controller_spec.rb]: user.update!...

File: spec/controllers/profile_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      @user.save!
+      get :edit
+      expect(response).to render_template('edit')
+    end
+  end
+
+  context 'not post to' do
```
Comment (RGBD) [inline spec/controllers/profile_controller_spec.rb]: Again, name is misleading

File: spec/controllers/profile_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    it 'update' do
+      sign_in @user
+      @user.role = :developer
+      @user.save!
+      post 'update', params: { phone_number: '3423432', company_name: 'company_name user' }
+      expect(response).to redirect_to(root_url)
```
Comment (RGBD) [inline spec/controllers/profile_controller_spec.rb]: also check that params were actually saved

File: spec/controllers/profile_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    end
+
+    it 'update role' do
+      sign_in @user
+      post 'update_role', params: { role: :customer }
+      expect(response).to redirect_to(customer_setup_info_profile_path.to_s.split('?').first)
```
Comment (RGBD) [inline spec/controllers/profile_controller_spec.rb]: Also check that params were actually saved

File: spec/controllers/projects_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,108 @@
+require 'rails_helper'
+
+RSpec.describe ProjectsController, type: :controller do
+  before(:each) do
+    @user = User.create!(name: 'midd', email: 'user@example.org', password: 'very-secret')
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: Maybe we can DRY user creation since it used many times in specs?

File: spec/controllers/projects_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    @user.role = :developer
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: Just add role: :developer to create!

File: spec/controllers/projects_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    @user.save!
+  end
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: newline

File: spec/controllers/projects_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+  context 'get' do
+    it 'returns a successful response' do
+      get :index
+      expect(response).to be_successful
+    end
+
+    it 'assigns @projects' do
+      @project = Project.create(title: 'Project1',
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: Can we move user and project creation in their own method?
Maybe it user creation would fit in before block?

File: spec/controllers/projects_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+                                info: 'body project',
+                                user: User.new(email: 'midda999@mail.ru',
+                                               name: 'Anton',
+                                               password: '123456'))
+      get :index
+      expect(assigns(:projects)).to eq([@project])
+    end
+
+    it 'project find by id' do
+      @project = Project.new(title: 'Project1', info: 'info to project 1', user: @user)
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: User.new && user.save! == User.create!

File: spec/controllers/projects_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+      @project.save!
+      get :show, params: { id: @project.id }
+      expect(assigns(:project)).to eq(@project)
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: Read about testing views in rails

File: spec/controllers/projects_controller_spec.rb [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+    it 'author' do
+      @project = Project.new(title: 'Project1', info: 'info to project 1', user: @user)
+      @project.save!
+      get :edit, params: { id: @user.projects.first }
+      expect(@user.projects.first).to eq(@project)
+    end
+
+    it 'not author' do
+      sign_in @user
+      user2 = User.create!(name: 'midd22', email: 'usetrr@example.org', password: 'ver12y-secret')
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: Again, can we unify user and project creation in before and do minimal setup inside unit tests?

File: spec/controllers/projects_controller_spec.rb [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+      project = Project.new(title: 'Project1', info: 'info for project 1', user: user2)
+      project.save!
+      user2.save!
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: Why do you need to save user2 here after creation in line 56?

File: spec/controllers/projects_controller_spec.rb [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+      get :edit, params: { id: user2.projects.first }
+      expect(@user).not_to eq(project.user)
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: What are you testing here exactly?
You can check that project has correct user in previous test.

File: spec/controllers/projects_controller_spec.rb [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+      get :new
+      expect(response).to render_template('new')
+    end
+  end
+
+  context 'project edit by user' do
+    it 'author' do
+      @project = Project.new(title: 'Project1', info: 'info to project 1', user: @user)
+      @project.save!
+      get :edit, params: { id: @user.projects.first }
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: Try removing this line. If test still passes, then you're not testing anything meaningful

File: spec/controllers/projects_controller_spec.rb [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+      project = Project.new(title: 'Project1', info: 'info for project 1', user: user2)
+      project.save!
+      user2.save!
+      get :edit, params: { id: user2.projects.first }
+      expect(@user).not_to eq(project.user)
+    end
+  end
+
+  context 'post' do
+    it 'create new project if true' do
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: misleading description, please use english language.

File: spec/controllers/projects_controller_spec.rb [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+      sign_in @user
+      project = Project.new(title: 'Project1', info: 'info for project 1', user: @user)
+      post :create, params: JSON.parse(project.to_json), as: :json
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: You don't need to create project. juse pass { title: 'Project', info: '...', ... } as params

File: spec/controllers/projects_controller_spec.rb [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+      expect(response).to redirect_to(projects_path)
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: also check that params were persisted.

File: spec/controllers/projects_controller_spec.rb [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+    end
+
+    it 'create new project if false' do
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: misleading name, use english language

File: spec/controllers/projects_controller_spec.rb [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+      sign_in @user
+      project = Project.new(title: nil, info: nil)
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: don't create project here, just pass params to post.

File: spec/controllers/projects_controller_spec.rb [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+      post :create, params: JSON.parse(project.to_json), as: :json
+      expect(response.location).to redirect_to new_project_url
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: check that params were persisted.
Seems like you tried to reach 100% coverage without actually checking anything.

File: spec/controllers/projects_controller_spec.rb [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+    end
+  end
+
+  context 'update project' do
+    it 'if all right' do
+      sign_in @user
+      project = Project.create(title: 'Project1', info: 'info for project 1', user: @user)
+      post :update, params: { id: project.id, project: JSON.parse(project.to_json), as: :json }
+      expect(response).to redirect_to(project_path)
```
Comment (RGBD) [inline spec/controllers/projects_controller_spec.rb]: Check that params were persisted.

File: spec/models/project_spec.rb [hunk 27]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Project, type: :model do
+  context 'can' do
```
Comment (RGBD) [inline spec/models/project_spec.rb]: why is context called can?

Comment (middaysan) [reply spec/models/project_spec.rb]: fix

File: spec/controllers/projects_controller_spec.rb (hunk 26 above)
Comment (middaysan) [reply spec/controllers/projects_controller_spec.rb]: done

File: spec/controllers/projects_controller_spec.rb (hunk 25 above)
Comment (middaysan) [reply spec/controllers/projects_controller_spec.rb]: done
why?  i check create method on correct redirect after create or not create object, yes, mistake what need check  params persisted

File: spec/controllers/projects_controller_spec.rb (hunk 24 above)
Comment (middaysan) [reply spec/controllers/projects_controller_spec.rb]: where i take project.id then?
image

File: app/controllers/projects_controller.rb (hunk 2 above)
Comment (middaysan) [reply app/controllers/projects_controller.rb]: undo change, done

File: spec/controllers/profile_controller_spec.rb (hunk 3 above)
Comment (middaysan) [reply spec/controllers/profile_controller_spec.rb]: done

File: spec/controllers/profile_controller_spec.rb (hunk 4 above)
Comment (middaysan) [reply spec/controllers/profile_controller_spec.rb]: rename, done

File: spec/controllers/profile_controller_spec.rb (hunk 5 above)
Comment (middaysan) [reply spec/controllers/profile_controller_spec.rb]: done

File: spec/controllers/profile_controller_spec.rb (hunk 6 above)
Comment (middaysan) [reply spec/controllers/profile_controller_spec.rb]: done

File: spec/controllers/profile_controller_spec.rb (hunk 7 above)
Comment (middaysan) [reply spec/controllers/profile_controller_spec.rb]: done

File: spec/controllers/profile_controller_spec.rb (hunk 8 above)
Comment (middaysan) [reply spec/controllers/profile_controller_spec.rb]: done

File: spec/controllers/profile_controller_spec.rb (hunk 9 above)
Comment (middaysan) [reply spec/controllers/profile_controller_spec.rb]: done

File: spec/controllers/projects_controller_spec.rb (hunk 11 above)
Comment (middaysan) [reply spec/controllers/projects_controller_spec.rb]: done

File: spec/controllers/projects_controller_spec.rb (hunk 12 above)
Comment (middaysan) [reply spec/controllers/projects_controller_spec.rb]: done

File: Gemfile (hunk 1 above)
Comment (middaysan) [reply Gemfile]: Remind me why did you need to install github version and not rubygems.org? - Now I know, 
don't do that

by the way, you saw this code in the camp and didn’t tell me anything) 

done

File: spec/controllers/projects_controller_spec.rb (hunk 14 above)
Comment (middaysan) [reply spec/controllers/projects_controller_spec.rb]: fix

File: spec/controllers/projects_controller_spec.rb (hunk 17 above)
Comment (middaysan) [reply spec/controllers/projects_controller_spec.rb]: delete and change new to create

File: spec/controllers/projects_controller_spec.rb (hunk 21 above)
Comment (middaysan) [reply spec/controllers/projects_controller_spec.rb]: project.id require in params

File: spec/controllers/projects_controller_spec.rb (hunk 22 above)
Comment (middaysan) [reply spec/controllers/projects_controller_spec.rb]: done

File: spec/controllers/projects_controller_spec.rb (hunk 19 above)
Comment (middaysan) [reply spec/controllers/projects_controller_spec.rb]: fix

File: spec/controllers/home_controller_spec.rb [hunk 28]
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
+
+RSpec.describe HomeController, type: :controller do
+  context 'not get' do
+    it 'index' do
+      get :index
+      expect(response).to render_template('index')
+    end
```
Comment (middaysan) [inline spec/controllers/home_controller_spec.rb]: add new project and check it on main page, test

File: spec/controllers/projects_controller_spec.rb [hunk 29]
```diff
@@ -0,0 +1,112 @@
...
+                         email: 'user@example.org',
+                         password: 'very-secret',
+                         role: :developer)
+    @user.save!
+  end
+
+  context 'get' do
+    it 'returns a successful response' do
+      get :index
+      expect(response).to be_successful
```
Comment (middaysan) [inline spec/controllers/projects_controller_spec.rb]: after get project check his params and description

File: spec/controllers/projects_controller_spec.rb (hunk 23 above)
Comment (middaysan) [reply spec/controllers/projects_controller_spec.rb]: project #create

File: spec/controllers/projects_controller_spec.rb [hunk 30]
```diff
@@ -0,0 +1,112 @@
...
+      expect(response).to redirect_to(project_path)
+      pr = Project.find_by_id(project.id)
+      expect(pr.title).to eq('Project1')
+    end
+
+    it 'if bad data project' do
+      sign_in @user
+      project = Project.create(title: 'Project1', info: 'info for project 1', user: @user)
+      project[:title] = ''
+      project[:info] = ''
```
Comment (middaysan) [inline spec/controllers/projects_controller_spec.rb]: replace wtih json

File: spec/controllers/projects_controller_spec.rb [hunk 31]
```diff
@@ -0,0 +1,112 @@
...
+      expect(project.title).to eq('Project1')
+    end
+  end
+
+  context 'destroy' do
+    it 'project' do
+      sign_in @user
+      project = Project.create(title: 'Project1', info: 'info for project 1', user: @user)
+      post :destroy, params: { id: project.id, project: JSON.parse(project.to_json), as: :json }
+      expect(response).to redirect_to(projects_path)
```
Comment (middaysan) [inline spec/controllers/projects_controller_spec.rb]: reload and test

File: spec/models/project_spec.rb [hunk 32]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Project, type: :model do
+  context 'Project can' do
+    it 'not create without user' do
+      project = Project.new(title: 'Best title', info: 'Best info')
+      expect(project.valid?).to eq(false)
```
Comment (middaysan) [inline spec/models/project_spec.rb]: expect errors, project.errors.messages - must have project

File: spec/models/project_spec.rb [hunk 33]
```diff
@@ -0,0 +1,20 @@
...
+    end
+    it 'create with user' do
+      project = Project.new(title: 'Best title', info: 'Best info', user: User.new)
+      expect(project.valid?).to eq(true)
+    end
+    it 'have skills' do
+      project = Project.new(skills: %i[css java ruby])
+      expect(project.skills).to eq(%w[css java ruby])
+    end
```
Comment (middaysan) [inline spec/models/project_spec.rb]: -

File: spec/models/user_spec.rb [hunk 34]
```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  context 'should not valid with' do
+    it 'wrong mail' do
+      user = User.new(email: 'midda', name: 'Anton', password: '123456')
+      expect(user.valid?).to eq(false)
```
Comment (middaysan) [inline spec/models/user_spec.rb]: expect user.errors.messages - specific messages

File: spec/models/user_spec.rb [hunk 35]
```diff
@@ -0,0 +1,62 @@
...
+    end
+    it 'empty password' do
+      user = User.new(email: 'midda999@mail.ru', name: 'Anton', password: '')
+      expect(user.valid?).to eq(false)
+    end
+  end
+
+  context 'can' do
+    it 'have project' do
+      user = User.new(email: 'midda999@mail.ru', name: 'Anton', password: '123456')
```
Comment (middaysan) [inline spec/models/user_spec.rb]: add before block

File: spec/models/user_spec.rb [hunk 36]
```diff
@@ -0,0 +1,62 @@
...
+      user.skills = :css
+      expect(user.skills).to eq('css')
+    end
+  end
+
+  context 'valid' do
+    it 'with right fields' do
+      user = User.new(email: 'midda999@mail.ru', name: 'Anton', password: '123456')
+      expect(user.valid?).to eq(true)
```
Comment (middaysan) [inline spec/models/user_spec.rb]: expect user.errors.messages

File: spec/models/user_spec.rb [hunk 37]
```diff
@@ -0,0 +1,62 @@
...
+    end
+    it 'omniauth' do
```
Comment (middaysan) [inline spec/models/user_spec.rb]: delete

File: spec/rails_helper.rb [hunk 38]
```diff
@@ -0,0 +1,64 @@
+# This file is copied to spec/ when you run 'rails generate rspec:install'
+require 'spec_helper'
+ENV['RAILS_ENV'] ||= 'test'
+
+require File.expand_path('../config/environment', __dir__)
+
+# Prevent database truncation if the environment is production
+abort("The Rails environment is running in production mode!") if Rails.env.production?
+require 'rspec/rails'
+# Add additional requires below this line. Rails is not loaded until this point!
```
Comment (middaysan) [inline spec/rails_helper.rb]: delete comments

File: spec/spec_helper.rb [hunk 39]
```diff
@@ -0,0 +1,99 @@
+# This file was generated by the `rails generate rspec:install` command. Conventionally, all
```
Comment (middaysan) [inline spec/spec_helper.rb]: delete comments

File: spec/views/home/index.html.erb_spec.rb [hunk 40]
```diff
@@ -0,0 +1,10 @@
+require 'spec_helper'
```
Comment (middaysan) [inline spec/views/home/index.html.erb_spec.rb]: delete check and this file

File: spec/controllers/profile_controller_spec.rb [hunk 41]
```diff
@@ -0,0 +1,48 @@
...
+    @user = User.create!(name: 'midd', email: 'user@example.org', password: 'very-secret')
+    sign_in @user
+  end
+
+  context 'Correct template response' do
+    before(:each) do
+      @user.update(role: :customer)
+    end
+
+    it 'edit role' do
```
Comment (middaysan) [inline spec/controllers/profile_controller_spec.rb]: move to separate context, check that set_role actually sets role

File: spec/controllers/projects_controller_spec.rb [hunk 42]
```diff
@@ -0,0 +1,91 @@
...
+  context 'project edit by user' do
+    it 'author' do
+      get :edit, params: { id: @project.id }
+      expect(response).to render_template('edit')
+    end
+
+    it 'not author' do
+      user2 = User.create(name: 'midd22', email: 'usetrr@example.org', password: 'ver12y-secret')
+      project2 = Project.create(title: 'Project1', info: 'info for project 1', user: user2)
+      get :edit, params: { id: user2.projects.first }
```
Comment (middaysan) [inline spec/controllers/projects_controller_spec.rb]: project2.id

File: spec/controllers/projects_controller_spec.rb [hunk 43]
```diff
@@ -0,0 +1,91 @@
...
+      expect(@user).not_to eq(project2.user)
```
Comment (middaysan) [inline spec/controllers/projects_controller_spec.rb]: check that updating another one user's project results in error

File: spec/controllers/projects_controller_spec.rb [hunk 44]
```diff
@@ -0,0 +1,91 @@
...
+    end
+  end
+
+  context 'project #create with' do
+    it 'correct params' do
+      post :create, params: JSON.parse(@project.to_json), as: :json
```
Comment (middaysan) [inline spec/controllers/projects_controller_spec.rb]: DO NOT USE JSON PARSE(.to_json)

File: spec/models/user_spec.rb [hunk 45]
```diff
@@ -0,0 +1,51 @@
...
+    end
+  end
+
+  context 'can' do
+    before(:each) do
+      @user = User.new(email: 'midda999@mail.ru', name: 'Anton', password: '123456')
+    end
+
+    it 'have role' do
+      @user.role = :customer
```
Comment (middaysan) [inline spec/models/user_spec.rb]: test doens't make sense

File: spec/models/user_spec.rb [hunk 46]
```diff
@@ -0,0 +1,51 @@
...
+      expect(@user.role).to eq('customer')
+    end
+
+    it 'have skills' do
+      expect(@user.skills).to eq([])
```
Comment (middaysan) [inline spec/models/user_spec.rb]: add user skills setter

File: app/models/user.rb [hunk 47]
```diff
@@ -34,4 +34,10 @@ def self.from_omniauth(auth)
       )
     end
   end
+
+  def skills=(args)
+    raise 'wrong input in array' unless args.is_a?(Array)
```
Comment (middaysan) [inline app/models/user.rb]: array expected

File: spec/controllers/profile_controller_spec.rb [hunk 48]
```diff
@@ -0,0 +1,48 @@
...
+    @user = User.create!(name: 'midd', email: 'user@example.org', password: 'very-secret')
+    sign_in @user
+  end
+
+  it 'Correct template response edit role' do
+    get :edit_role
+    expect(response).to render_template('edit_role')
+  end
+
+  context 'Correct template response' do
```
Comment (middaysan) [inline spec/controllers/profile_controller_spec.rb]: testing customer

File: spec/controllers/profile_controller_spec.rb [hunk 49]
```diff
@@ -0,0 +1,48 @@
...
+      expect(response).to render_template('customer_setup_info')
+    end
+
+    it 'edit' do
+      get :edit
+      expect(response).to render_template('edit')
+    end
+  end
+
+  context 'Correct redirect after post' do
```
Comment (middaysan) [inline spec/controllers/profile_controller_spec.rb]: testing developer

File: spec/controllers/projects_controller_spec.rb (hunk 22 above)
Comment (middaysan) [reply spec/controllers/projects_controller_spec.rb]: NOT DONE!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!1111111111111111!!!!!!!!!!!!!!!!!!!!!!!!!!!11111!!!!!!!111!!!!!!!!!!!!!ONE

File: spec/controllers/profile_controller_spec.rb [hunk 50]
```diff
@@ -0,0 +1,48 @@
...
+    it 'update' do
+      @user.update(role: :developer)
+      post 'update', params: { phone_number: '3423432', company_name: 'company_name user' }
+      expect(response).to redirect_to(root_url)
+      @user.reload
+      expect(@user.phone_number).to eq('3423432')
+      expect(@user.company_name).to eq('company_name user')
+    end
+
+    it 'update role' do
```
Comment (middaysan) [inline spec/controllers/profile_controller_spec.rb]: wrong context, must be customer

File: spec/controllers/projects_controller_spec.rb [hunk 51]
```diff
@@ -0,0 +1,95 @@
...
+      expect(response).to render_template('edit')
+    end
+
+    it 'not author' do
+      user2 = User.create(name: 'midd22', email: 'usetrr@example.org', password: 'ver12y-secret')
+      project2 = Project.create(title: 'Project1', info: 'info for project 1', user: user2)
+      expect do
+        post :update, params: { id: project2.id, project: { title: 'Get error',
+                                                            info: 'body error',
+                                                            user: @user } }
```
Comment (middaysan) [inline spec/controllers/projects_controller_spec.rb]: remove line

File: spec/controllers/projects_controller_spec.rb [hunk 52]
```diff
@@ -0,0 +1,95 @@
...
+  context 'project #create with' do
+    it 'correct params' do
+      post :create, params: { project: { title: 'New project to create', info: 'body for create' } }
+      expect(response).to redirect_to(projects_path)
+      expect(Project.last.title).to eq('New project to create')
+    end
+
+    it 'wrong params' do
+      expect do
+        post :create, params: { project: { title: nil, info: nil, user: @user } }
```
Comment (middaysan) [inline spec/controllers/projects_controller_spec.rb]: remove user: user

File: spec/models/project_spec.rb [hunk 53]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Project, type: :model do
+  context 'Project can' do
+    it 'not create without user' do
+      project = Project.new(title: 'Best title', info: 'Best info')
+      expect(project.valid?).to eq(false)
+      expect(project.errors.messages).to eq(user: ['must exist'])
```
Comment (middaysan) [inline spec/models/project_spec.rb]: project.errors.messages[:user].to include('must exist')

File: spec/models/user_spec.rb [hunk 54]
```diff
@@ -0,0 +1,48 @@
...
+
+  context 'can' do
+    before(:each) do
+      @user = User.new(email: 'midda999@mail.ru', name: 'Anton', password: '123456')
+    end
+
+    it 'have skills' do
+      expect { @user.skills = '' }.to raise_error(RuntimeError)
+      @user.skills = %w[1 2]
+      expect(@user.skills.is_a?(Array)).to eq(true)
```
Comment (middaysan) [inline spec/models/user_spec.rb]: user.skills.to eq(%w[1, 2])

File: spec/controllers/projects_controller_spec.rb [hunk 55]
```diff
@@ -0,0 +1,94 @@
...
+                                                            info: 'body error' } }
+      end.to raise_error(CanCan::AccessDenied)
+    end
+  end
+
+  context 'project #create with' do
+    it 'correct params' do
+      post :create, params: { project: { title: 'New project to create', info: 'body for create' } }
+      expect(response).to redirect_to(projects_path)
+      expect(Project.last.title).to eq('New project to create')
```
Comment (middaysan) [inline spec/controllers/projects_controller_spec.rb]: project.exissts(title:..., info: ...)

File: spec/controllers/projects_controller_spec.rb [hunk 56]
```diff
@@ -0,0 +1,116 @@
...
+  context 'project edit by user' do
+    it 'author' do
+      get :edit, params: { id: @project.id }
+      expect(response).to render_template('edit')
+    end
+
+    it 'not author' do
+      user2 = User.create(name: 'midd22', email: 'usetrr@example.org', password: 'ver12y-secret', role: 'developer')
+      project2 = Project.create(
+          title: 'Project12',
```
Comment (middaysan) [inline spec/controllers/projects_controller_spec.rb]: use two spaces for indentation

File: spec/controllers/projects_controller_spec.rb [hunk 57]
```diff
@@ -0,0 +1,116 @@
...
+        post :update, params: { id: project2.id, project: { title: 'Get error',
+                                                            info: 'body error' } }
+      end.to raise_error(CanCan::AccessDenied)
+    end
+  end
+
+  context 'project #create with' do
+    it 'correct params' do
+      project_params = {
+          title: 'Project1',
```
Comment (middaysan) [inline spec/controllers/projects_controller_spec.rb]: indent
